### PR TITLE
fix(python): reuse state in `to_pyarrow_dataset`

### PR DESCRIPTION
**Pull request by ion-elgreco**
# Description
Reuse the state so we save time instead of reconstructing and verifying the state.

### CODE REVIEW

**Comment by adriangb** [comment]
Yup this fixes the issue 🥳

**Comment by ion-elgreco** [comment]
> @ion-elgreco looks like it needs a rebase. Also just curious if you'll be making a release with this fix since it's such a huge performance win for an essential feature?

@wjones127 could you take a look at the PR?

File: python/tests/test_fs.py [hunk 1]
```diff
@@ -241,16 +240,3 @@ def test_roundtrip_azure_decoded_sas(azurite_sas_creds, sample_data: pa.Table):
...
-    store = DeltaStorageHandler(str(tmp_path.absolute()))
-
-    with (tmp_path / "asd.pkl").open("wb") as handle:
-        pickle.dump(store, handle)
-
-    with (tmp_path / "asd.pkl").open("rb") as handle:
-        store_pkl = pickle.load(handle)
-
-    infos = store_pkl.get_file_info(["asd.pkl"])
-    assert infos[0].size > 0
```
Comment (wjones127) [inline python/tests/test_fs.py:256]: Are we dropping the ability to pickle?

File: python/src/filesystem.rs [hunk 2]
```diff
@@ -41,21 +41,27 @@ impl DeltaFileSystemHandler {
 #[pymethods]
 impl DeltaFileSystemHandler {
     #[new]
-    #[pyo3(signature = (table_uri, options = None, known_sizes = None))]
+    #[pyo3(signature = (table_uri, table = None, options = None, known_sizes = None))]
     fn new(
-        table_uri: &str,
+        table_uri: String,
+        table: Option<&RawDeltaTable>,
```
Comment (wjones127) [inline python/src/filesystem.rs:47]: Maybe instead make a separate classmethod `from_table` that takes the table? Then perhaps we can keep the pickling support.

**Review by wjones127 — APPROVED** [review]
Looks better now. Thanks.

**Pull request merged**